src/irmin/proof.ml

```diff
@@ -478,18 +503,34 @@ struct
                 bad_stream_exn "find_node"
                   "empty stream when looking for hash %a" pp_hash h
             | Cons (v, rest) ->
-                (* [depth] is 0 because this context deals with root nodes *)
-                let v = rehydrate_stream_node ~depth:0 v h in
-                let v = N.with_handler (find_recnode t) v in
-                check_node_integrity v h;
+                (* Shorten [stream] before calling [head] as it might itself
+                   perform reads. *)
                 stream := rest;
+                let v =
+                  (* [depth] is 0 because this context deals with root nodes *)
+                  rehydrate_stream_node ~depth:0 v h
+                in
+                let v =
+                  (* Call [with_handler] before [head] because the later might
+                     perform reads *)
+                  N.with_handler (find_recnode t) v
+                in
+                let _ =
```

Comment:
```suggestion
               let _ : [ `Node of _ | `Inode of _ ] =
```

(Not sure why `N.head` is a local substitution when `N.proof` is provided as a type.)